### DISCUSSION TITLE
Improved "Privacy" Options: Alternative to Google Play on Site, In-App Offline Mode?

### DISCUSSION THREAD

**Discussion by mochsner** [ezra-bible-app/ezra-bible-app#971] [Ideas]
Not sure if this is the right place for this (versus discussions), but as a new user (just discovered Ezra today), the first thing I saw (on Android), was the big red error "IMPORTANT NOTE: If you are located in a country where Christians are persecuted, be aware that your internet provider will be able to detect that **you are accessing Christian resources!**. 

And, well, it just made me think about how one might be able to circumvent this, as it seems like a great app and in my opinion seems like it will long-term have more adoption over BibleTime (which I just came from). 

Some thoughts on ta

(1) There's an Android .apk on Github, but it's nowhere on the website (which can be translated with a guide on how to install). Could this possibly be advertised/documented there and/or else hosted on f-droid which doesn't have auto-update capabilities? I would NOT trust Google with protecting or withholding any data from governments, as their objective is entirely to make money. 

(2) Since we use SQLite on Cordova within the app, how feasible would it be to have an "offline" mode where nothing is pushed, pulled, etc. from any servers? That way, users might be able to use a VPN to initially connect, download resources, etc - but then disconnect and not have to worry so much.

**Comment by tobias-klein** [comment]
Hi @mochsner

Thanks for your feedback!
Resuming this issue as a discussion as it addresses multiple topics and we may need to discuss a bit before turning it into an issue that could be worked on.

1) I only started recently to upload the APK (based on #919). We could also link it on the Website, even though I think that a direct installation via APK is probably for advanced users / specialist people. Also, I need to check out F-Droid one of these days ... unfortunately the build environment for Android is kind of tricky based on the set of technologies I'm using and therefore much of the release process is manual. That means introducing F-Droid may add more work for each release ... anyway, I'll have a look one of these days.
2) Technically the app is already working offline (mostly). The internet connection is currently needed to:
a) Download SWORD modules (can only be avoided by pre-packaging SWORD modules with the app). Once they are downloaded, though - there is no additional internet access needed.
b) Check on the latest version available (GitHub).
c) Upload crash reports to Sentry (a diagnostics platform used to debug issues).
d) Synchronize with Dropbox.

Point 2b) and 2c) could be made configurable. Point 2d) is anyway only activated based on enabling it first.

So, as I mentioned above - for the most part the app already functions offline and if you want to have a complete offline behavior we could make 2b) and 2c) configurable.

**Reply by tobias-klein** [reply]
@mochsner
I have created issues #988 and #989 now based on 2b) and 2c) mentioned above. Both will be fixed in the next release.

**Comment by sosukeinu** [comment]
You mentioned that the only way to avoid the initial Internet connection is to pre-package sword modules with the app. I think even this can be circumvented by allowing an installation from a .zip file. Another Android Sword Project app called Bishop allows for this, but it lacks many of the advanced features available in Ezra such as the variant viewer, etc. What are your thoughts about inclusion of an "install module from .zip," feature? This would allow those in persecuted countries to pass modules around in multiple ways, and install and stay offline.

**Reply by tobias-klein** [reply]
@sosukeinu Yes, that could probably be done. We would have to come up with a UI for that, though.
Since this is a bit of a rare usecase (not many people requested this so far), we could alternatively just provide documentation on how to put SWORD modules into the right folder on the various platforms. On Android this should be `/sdcard/Documents/sword` (currently not implemented, though).

If a certain path is read by default, then the user would just have to place the modules in that folder (outside of Ezra) and Ezra would read any modules available from that folder. What do you think about that? With that approach we could skip implementing a separate UI just for the "offline" installation.

**Comment by sosukeinu** [comment]
I actually prefer that implementation, as the less friction the better.
Documentation would be key with an implementation like this.
Thank you for the fast response.

**Reply by tobias-klein** [reply]
You're welcome! I may consider adding some built-in documentation for this approach in one of the next versions. I just created a new issue for that (see #1000).
